Kafka/src/main/java/io/deephaven/kafka/KafkaTools.java

```diff
@@ -661,28 +679,39 @@ DataFormat dataFormat() {
-                    if (excludeColumns == null && includeColumns != null) {
-                        return includeColumns;
+                    if (excludeColumns != null && includeColumns != null) {
+                        throw new IllegalArgumentException("Can't have both excludeColumns and includeColumns not null");
                     }
                     final String[] tableColumnNames = t.getDefinition().getColumnNamesArray();
-                    final String[] mismatches = Arrays.stream(tableColumnNames)
-                            .filter(cn -> !excludeColumns.contains(cn)).toArray(String[]::new);
-                    if (mismatches.length > 0) {
-                        throw new IllegalArgumentException(
-                                "excludeColumns contains column names not on the table: " + mismatches);
+                    if (excludeColumns == null && includeColumns == null) {
+                        return tableColumnNames;
                     }
-                    final ColumnDefinition<?>[] colDefs = t.getDefinition().getColumns();
-                    final int nColNames = colDefs.length - excludeColumns.size();
-                    final String[] colNames = new String[nColNames];
-                    int c = 0;
-                    for (final String colName : tableColumnNames) {
-                        if (excludeColumns != null && excludeColumns.contains(colName)) {
-                            continue;
+                    final Set<String> tableColumnsSet = new HashSet<>(Arrays.asList(tableColumnNames));
+                    if (includeColumns != null) {
+                        // Validate includes
+                        final List<String> missing = Arrays.stream(includeColumns).
+                                filter(cn -> !tableColumnsSet.contains(cn)).
+                                collect(Collectors.toList());
+                        for (final String col : includeColumns) {
+                            if (!tableColumnsSet.contains(col)) {
+                                missing.add(col);
+                            }
+                        }
+                        if (missing.size() > 0) {
+                            throw new IllegalArgumentException(
+                                    "includeColumns contains names not found in table columns: " + missing);
                         }
-                        colNames[c++] = colName;
+                        return includeColumns;
                     }
-
-                    return colNames;
+                    final List<String> mismatches = excludeColumns.stream().
+                            filter(cn -> !tableColumnsSet.contains(cn)).
+                            collect(Collectors.toList());
+                    if (mismatches.size() > 0) {
+                        throw new IllegalArgumentException(
+                                "excludeColumns contains column names not on the table: " + mismatches);
```

Comment:
Maybe word this the same as the one above?
`"excludeColumns contains names not found in table columns: "`